org.eclipse.scout.rt.mail/src/main/java/org/eclipse/scout/rt/mail/smtp/SmtpConnectionPool.java

```diff
@@ -158,22 +168,37 @@ public void sendMessage(SmtpServerConfig smtpServerConfig, MimeMessage message,
    */
   protected SmtpConnectionPoolEntry leaseConnection(SmtpServerConfig smtpServerConfig) {
     Assertions.assertGreater(smtpServerConfig.getPoolSize(), 0, "Pool size of provided SmtpServerConfig must be greater 0.");
-    synchronized (m_poolLock) {
-      Assertions.assertFalse(m_destroyed, "SmtpConnectionPool not available because it has already been destroyed.");
-      SmtpConnectionPoolEntry candidate = null;
-      while (candidate == null) {
+    while (true) {
+      boolean createNewConnection = false;
+      synchronized (m_poolLock) {
+        checkAndThrowIfDestroyed();
         // try to find an idle connection or create a new connection if poolsize has not been reached yet.
         // candidate may be null as a result
-        try {
-          candidate = tryGetIdleOrCreateNewConnection(smtpServerConfig);
-        }
-        catch (RuntimeException | MessagingException e) {
-          throw new ProcessingException("MessagingException caught while trying to connect to smtp server.", e);
+        SmtpConnectionPoolEntry candidate = tryGetIdleConnection(smtpServerConfig);
+        if (candidate != null) {
+          // early return non-null candidate
+          m_idleEntries.remove(candidate);
+          m_leasedEntries.add(candidate);
+          LOG.debug("Leasing pooled SMTP connection {}", candidate);
+          return candidate;
         }
 
+        Predicate<SmtpConnectionPoolEntry> poolFilter = getPoolFilter(smtpServerConfig);
+        // there was no idle connection so we create a new connection if the pool size has not been reached for this config
+        long idleEntryCount = m_idleEntries.stream()
+            .filter(poolFilter)
+            .count();
+        long leasedEntryCount = m_leasedEntries.stream()
+            .filter(poolFilter)
+            .count();
+        int currentlyConnectingCount = m_currentlyConnectingCounter.compute(smtpServerConfig, (k, v) -> v == null ? 1 : (v + 1)); // increment
+        createNewConnection = idleEntryCount + leasedEntryCount + currentlyConnectingCount <= smtpServerConfig.getPoolSize();
+
         // if we could not find an idle connection and the pool has already reached its limit in terms of connection count,
         // we wait until someone releases a connection (@see #releaseConnection(SmtpConnectionPoolEntry))
-        if (candidate == null) {
+        if (!createNewConnection) {
+          m_currentlyConnectingCounter.compute(smtpServerConfig, (k, v) -> v == 1 ? null : (v - 1)); // decrement (remove if 0)
```

Comment:
try-finally over whole block to decrement? currently there is no code in between that could fail and thus counter is not decrement, but might be added unintentionally later

Reply:
Not sure what would be the best way to write this try-finally block, maybe:

```java
        try {
          createNewConnection = idleEntryCount + leasedEntryCount + currentlyConnectingCount <= smtpServerConfig.getPoolSize();
        }
        finally {
          // if we could not find an idle connection and the pool has already reached its limit in terms of connection count,
          // we wait until someone releases a connection (@see #releaseConnection(SmtpConnectionPoolEntry))
          if (!createNewConnection) {
            decrementCurrentlyConnectingCounterSafe(smtpServerConfig);
          }
        }

        if (!createNewConnection) {
```

However now there would be a duplicate `if (!createNewConnection) {` block.

Reply:
As discussed, use try-finally and add an detailed comment explaining that after increment it must start with a try, in any case